src/provider/index.ts

```diff
@@ -103,32 +108,50 @@ export class Main {
     }
 
     private async open(payload: OpenPayload): Promise<void> {
+        const context = payload.context && parseContext(payload.context);
+
         const appInfo: Application|null = await this._directory.getAppByName(payload.name);
 
         if (!appInfo) {
             throw new FDC3Error(OpenError.AppNotFound, `No app in directory with name: ${payload.name}`);
         }
 
-        // This can throw FDC3Errors if app fails to open or times out
-        await this._model.ensureRunning(appInfo);
+        const promises: Promise<void>[] = [];
+
+        // Start the application if not already running
+        const {startedPromise} = await this._model.getOrCreateLiveApp(appInfo);
+
+        promises.push(startedPromise);
+
+        // If the app has open windows, bring all to front in creation order
+        const windows = this._model.findWindowsByAppName(appInfo.name);
+        if (windows.length > 0) {
+            windows.sort((a, b) => a.appWindowNumber - b.appWindowNumber);
+
+            const bringToFrontPromise = Promise.all(windows.map(window => window.bringToFront()));
+            const focusPromise = bringToFrontPromise.then(async () => {
+                await windows[windows.length - 1].focus();
```

Comment:
I think you can remove the `async` and `await` here and instead only return `windows[windows.length - 1].focus()`. The promise.all at the end should then await the focus resolve instead of the resolve of the focus resolve. This same logic can be applied to line 148 as well

Reply:
✅